// @jridgewell/gen-mapping/src/types.ts
import type { SourceMapSegment } from './sourcemap-segment';

export interface SourceMapV3 {
  file?: string | null;
  names: Array<string | null>;
  sourceRoot?: string;
  sources: Array<string | null>;
  sourcesContent?: Array<string | null>;
  version: 3;
}

export interface EncodedSourceMap extends SourceMapV3 {
  mappings: string;
}

export interface DecodedSourceMap extends SourceMapV3 {
  mappings: Array<Array<SourceMapSegment>>;
}

export interface Pos {
  line: number;
  column: number;
  sourceMap?: SourceMapV3;
}

export type Original = {
  line: number;
  column: number;
}

export type Mapping =
  | {
      generated: Pos;
      source: undefined;
      original: undefined;
      name: undefined;
    }
  | {
      generated: Pos;
      source: string;
      original: Original;
      name: string;
    }
  | {
      generated: Pos;
      source: string;
      original: Original;
      name: undefined;
    };

export type SourceMapSegment = {
  source: string;
  names: Array<string | null>;
}
